cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+            continue;
+        }
+
+        check_overlap_groups(&mut errors, &env, term);
+    }
+
+    if !errors.is_empty() {
+        let mut errors = errors.report(&env);
+        return match errors.len() {
+            1 => Err(errors.pop().unwrap()),
+            _ => Err(Error::Errors(std::mem::take(&mut errors))),
+        };
+    }
+
+    Ok(())
+}
+
+/// A node in the error graph.
+struct Node {
+    /// The number of other rules this node overlaps with.
+    degree: usize,
+
+    /// `true` entries where an edge exists to the node at that index.
+    edges: Vec<bool>,
+}
+
+impl Node {
+    /// Make a new `Node` in the error graph with the given number of total nodes.
+    fn new(len: usize) -> Self {
+        let mut edges = Vec::with_capacity(len);
+        edges.resize(len, false);
+        Self { degree: 0, edges }
+    }
+
+    /// Add an edge between this node and the other node.
+    fn add_edge(&mut self, other: RuleId) {
+        if self.edges[other.0] {
+            return;
+        }
+
+        self.degree += 1;
+        self.edges[other.0] = true;
+    }
+
+    /// Remove an edge between this node and another node.
+    fn remove_edge(&mut self, other: RuleId) {
+        if !self.edges[other.0] {
+            return;
+        }
+
+        self.degree -= 1;
+        self.edges[other.0] = false;
+    }
+}
+
+/// A graph of all the rules in the isle source, with bi-directional edges between rules that are
+/// discovered to have overlap problems.
+struct Errors {
+    /// Edges between rules indicating overlap. As the edges are not directed, the edges are
+    /// normalized by ordering the rule ids.
+    nodes: Vec<Node>,
+}
+
+impl Errors {
+    /// Make a new `Errors` graph for collecting overlap information.
+    fn new(len: usize) -> Self {
+        let mut nodes = Vec::with_capacity(len);
+        nodes.resize_with(len, || Node::new(len));
+        Self { nodes }
+    }
+
+    /// True when there are no edges in the graph.
+    fn is_empty(&self) -> bool {
+        self.nodes.iter().all(|node| node.degree == 0)
+    }
+
+    /// Condense the overlap information down into individual errors.
+    fn report(&mut self, env: &Env) -> Vec<Error> {
+        let mut rules: Vec<usize> = self
+            .nodes
+            .iter()
+            .enumerate()
+            .filter_map(|(id, node)| if node.degree == 0 { None } else { Some(id) })
+            .collect();
+
+        rules.sort_by_cached_key(|id| self.nodes[*id].degree);
+
+        let mut errors = Vec::new();
+
+        let get_info = |id| {
+            let rule = env.get_rule(id);
+
+            let src = env.get_source(rule.pos.file);
+            let span = Span::new_single(rule.pos);
+            (src, span)
+        };
+
+        // Work backwards through the ids to find the nodes with the largest conflict first.
+        for id in rules.into_iter().rev() {
+            let node = self.remove_edges(RuleId(id));
```

Comment:
Does this guarantee that the degree is monotonically decreasing? I can't quite pin it down but I feel like there's some situation where higher-degree rules can reduce some other rule's degree to the point that by the time you get there, it has lower degree than later rules (but still non-zero degree).

Here's someplace I'd actually be happier with an O(n^2) algorithm: Iterate over the rules to find one that currently has maximum degree (perhaps with `Iterator::max_by_key`), process it, and repeat. As long as `n` here is the number of rules that actually have overlaps, rather than the total number of rules, a quadratic algorithm is free in the expected case where we've already fixed all the overlaps.